GUIDE.md

```diff
@@ -0,0 +1,302 @@
+
+A predicate handler must implement `Handler` trait, and return
+a [`PredicateResult`](https://tg-rs.github.io/carapax/carapax/enum.PredicateResult.html)
+or a type that converts into it:
+
+|   From            | To                            |
+|-------------------|-------------------------------|
+| `true`            | `PredicateResult::Continue`   |
+| `false`           | `PredicateResult::Stop`       |
+| `Result<T, E>`    | `T.into::<PredicateResult>()` |
+
+
+Example:
+
+```rust
+use carapax::{
+    methods::SendMessage,
+    types::{ChatId, Text},
+    Api, Chain, PredicateExt, Ref,
+};
+
+fn setup(chain: &mut Chain) {
+    let handler = pong.predicate(is_ping);
+    chain.add_handler(handler);
+}
+
+async fn is_ping(text: Text) -> bool {
+    text.data == "ping"
+}
+
+async fn pong(api: Ref<Api>, chat_id: ChatId) {
+    let method = SendMessage::new(chat_id, "pong");
+    api.execute(method).await.unwrap();
+}
+
+```
+
+### Commands
+
+[`CommandPredicate`](https://tg-rs.github.io/carapax/carapax/struct.CommandPredicate.html#)
+is a decorator which allows to run a handler only if update contains a command.
+Note that command name contains a leading slash (`/`).
+
+```rust
+use carapax::{
+    methods::SendMessage,
+    types::{ChatId, User},
+    Api, Chain, CommandExt, Ref,
+};
+
+fn setup(chain: &mut Chain) {
+    let handler = greet.command("/hello");
+    chain.add_handler(handler);
+}
+
+async fn greet(api: Ref<Api>, chat_id: ChatId, user: User) {
+    let method = SendMessage::new(chat_id, format!("Hello, {}", user.first_name));
+    api.execute(method).await.unwrap();
+}
+
+```
+
+### Access
+
+[AccessPredicate](https://tg-rs.github.io/carapax/carapax/access/struct.AccessPredicate.html) allows to protect handlers.
+It takes an [`AccessPolicy`](https://tg-rs.github.io/carapax/carapax/access/trait.AccessPolicy.html).
+Policy has `is_granted` method which takes a `HandlerInput` and returns a future with `bool` output.
+If `true` is returned - access is granted, `false` - forbidden.
+
+[InMemoryAccessPolicy](https://tg-rs.github.io/carapax/carapax/access/struct.InMemoryAccessPolicy.html) is a policy
+which stores [access rules](https://tg-rs.github.io/carapax/carapax/access/struct.AccessRule.html) in memory.
+
+Let's see how it works:
+
+```rust
+use carapax::{
+    access::{AccessExt, AccessPredicate, AccessRule, InMemoryAccessPolicy},
+    types::Update,
+    Chain, Predicate, HandlerResult,
+};
+
+fn setup(chain: &mut Chain) {
+    let policy = InMemoryAccessPolicy::from(vec![AccessRule::allow_user("username")]);
+    let handler = Predicate::new(AccessPredicate::new(policy), protected_handler);
+    // or using AccessExt
+    // let handler = protected_handler.access(policy);
+    chain.add_handler(handler);
+}
+
+async fn protected_handler(update: Update) {
+    log::info!("Got a new update in protected handler: {:?}", update);
+}
+
+```
+
+Since `Chain` implements `Handler` you can also protect a group of handlers or entire bot.
```

Comment:
We cannot protect entire bot, we can protect the top-level handler (which can be chain, for example)